rfc/merkle-structure.md

```diff
@@ -0,0 +1,320 @@
+
+### Data Addressing
+
+Another side effect of having both local and external references is that it affects data addressing. This is easier explained with an example so let us consider following structure with a nested substructure under name `message`:
+
+```json
+{
+  "message": {
+    "from": "@gazala",
+    "to": "@mikeal",
+    "payload": "Hi"
+  }
+}
+```
+
+We could link to the `message` e.g. by creating [DAG-JSON] in which case it will have an address `baguqeeraljex7usivynmla6e6h2ah5otegcrxgcc3kturfknn26j3wf6xsjq` and the outer data structure will have address `baguqeerade2eyspg52ffd26o2obbrhaa6ztrnszcgosutmeblrr7cbrjcpfa`. If `message` is not externalized however, data structure will end up with an address `baguqeerade2eyspg52ffd26o2obbrhaa6ztrnszcgosutmeblrr7cbrjcpfa`.
+
+And this is just one dimension of variability, if we were to use [DAG-CBOR] encoding instead we would end up with `bafyreie7pbuqqem5fvmrh4wv4joy737tmgmoo5tn4q7qwocd2a52si2m6m` when `message` is externalized and with `bafyreie7pbuqqem5fvmrh4wv4joy737tmgmoo5tn4q7qwocd2a52si2m6m` otherwise.
+
+And this is not the end, we could have a case where `message` is in [DAG-JSON] and rest is in [DAG-CBOR] in which case our data will be addressed `bafyreig5bmydxjpbq5j46hqs5cnay7xs4yef2nu4zts4rdmgxjzlsh4ume`.
+
+ℹ️ IPLD shifted question from where content is to how is it encoded. It is better problem to have, but still a problem because how you want to encode it will depend on deployed system and undoubtedly will vary from case to case basis. It is also very likely that choice of encoding will change over time and change in addresses is a major drawback.
+
+## Proposition
+
+Now that we have outlined problems lets enumerate what would we desire instead:
+
+1. Data structure to be addressed by its contents regardless of location, whether some parts are inlined (local) or external. This will give us flexibility to decide which parts to inline vs externalize when storing and decide differently when transferring it.
+2. Data structure to be addressed by its content regardless of the encoding. This will give us flexibility to choose different encoding on disk from encoding on wire. In fact, we gain ability to negotiate content encoding per data exchange.
+
+### Proposed Addressing
+
+IPLD data model offers same data types as pretty much every mainstream programming language. We can discard link type due to issues it introduces. We also drop `null` because it is a bad idea and in most cases custom representation of "absence of value" is better choice.
```

Comment:
A single canonical representation that Merklizes _everything_, and has well defined Merklization of the data model types is a Very Good Idea:tm:

However, if I'm reading this correctly, you're saying that instead of an explicit link, we can a use can then break up the structure arbitrarily. I like that idea in general, but doesn't it mean that _any part of the structure_ can now have cache locality concerns? There's no way to express "if you have `a` you also definitely have `b`", right?

Hashes are very helpful in distributed systems, to the point that some top-tier DS researcher are working on a hash-based distributed operating system. Sure, everything is wrapped in a promise, requests can fail, time out, etc etc: that's how all DS work, and actor systems like Erlang & Elixir have been very successful with this model (using internal process IDs instead of CIDs, but close enough for this discussion).

I want to maintain structural sharing, both for usual data structure reasons, _but especially for caching._ For example: I cannot send the entire UCAN chain every time; I _must_ be able to omit parts of that payload. Being that it's signed over, I can't arbitrarily change the structure to suit serialisation needs.

# General Commentary

> We also drop `null` because it is a bad idea and in most cases custom representation of "absence of value" is better choice.

I'm not 100% sure that I agree. Yes, there's the old adage about the billion dollar mistake, but as you note: omitting a field is semantically equivalent (it's `undefined` by analogy to JS, which is arguably _worse_).

Absence is not always the same as marking something as empty. Empty strings, empty arrays, and empty maps are all examples of other valid "none" types that cause headaches (e.g. see the `NonEmpty` list type in Haskell and Rust). Falling back to hacks like `{field: ""}` to mean `None` seems like an unnecessary loss of expressive power.

Since this is a flat serialisation format, there probably isn't a requirement to have special serialisation for coproducts (which parser from this format to the application's type would do the conversion on), so having a `null` fits the role of `None` on required fields.

Reply:
> However, if I'm reading this correctly, you're saying that instead of an explicit link, we can a use can then break up the structure arbitrarily.

Idea is data is broken up completely all the way to the scalar level. Whether you capture all the hashes or just parts you care about is up to you and in that sense it is arbitrary.

> I like that idea in general, but doesn't it mean that any part of the structure can now have cache locality concerns? There's no way to express "if you have a you also definitely have b", right?

That is accurate and a fare concern. I need to think more about it. My immediate reaction is that belongs in a different layer.

> I want to maintain structural sharing, both for usual data structure reasons, but especially for caching. For example: I cannot send the entire UCAN chain every time; I must be able to omit parts of that payload. Being that it's signed over, I can't arbitrarily change the structure to suit serialisation needs.

I'm not sure I follow what you are trying to say here. I agree with what you are saying and I believe proposal here should fully support it, do you suggest otherwise ?

> No null

I'm amendable to adding null type, even though I personally feel that using custom `None` structure is generally a better idea.

Reply:
I have added `Null`. Even if I think it's better to model your data differently, in practice many languages use and have `null` and forcing them to not to is not an objective here.